### PR TITLE
feat!: Use the hugr builder

### LINKED ISSUE #85: `hugr.serialize` shouldn't mutate
Currently, `serialize` removes dummy nodes and adds state edges before doing the serialisation.
https://github.com/CQCL/guppylang/blob/89edc2d724586d20438395b2c9f141821753dabe/guppylang/hugr_builder/hugr.py#L783

This is unexpected behaviour. It should probably copy the Hugr before doing any mutation.

### LINKED ISSUE #257: refactor: port HUGR codegen to use HUGR python builder
~Blocked by non-optional tasks in tracking issue: https://github.com/CQCL/hugr/issues/486~

~Now blocked by https://github.com/CQCL/guppylang/issues/280 to avoid bad merge conflicts. (Base on working branch for that issue if starting before that is merged)~

### PR DESCRIPTION
**Pull request by aborgna-q**
This is an in-progress transplant of guppy from using an ad-hoc `Hugr` definition into the builder from the `hugr` library.

The main mismatch between the old and new builders is that guppy used to lazily add typed ports to the graph nodes as it needed. Since the new builder requires operations to define their signature, we end up having to edit the code everywhere from `guppylang.compiler` to the `guppylang.prelude` definitions.

My first goal is to get this working, and then I will try and split some changes from this PR (although most of it will have to be merged monolithically).
 
Current test status:
- 390 $\color{green}\text{passed}$
- 14 $\color{yellow}\text{skipped}$
- 0 $\color{red}\text{failed}$

These issues currently breake some tests. I'll fix them before merging the PR
- https://github.com/CQCL/hugr/issues/1319
  ~~Required to store the module names in the hugr metadata~~
- https://github.com/CQCL/hugr/issues/1424
  ~~Required for pytket and llvm integration~~
- ~~Release hugr and tket2 with the latest changes~~

Closes #257. Closes #85

BREAKING CHANGE: Removed `guppylang.hugr_builder.hugr.Hugr`, compiling a module returns a `hugr.ext.Package` instead. Use `module.compile_hugr` to obtain a `hugr.Hugr`.

BEGIN_COMMIT_OVERRIDE
feat!: Use the hugr builder

BREAKING CHANGE: Removed `guppylang.hugr_builder.hugr.Hugr`, compiling a module returns a `hugr.Package` instead.
END_COMMIT_OVERRIDE

### CODE REVIEW

File: guppylang/error.py
Comment (aborgna-q) [inline guppylang/error.py:1]: This is a drive-by change to avoid having the jupyter lookup error in the stack

File: guppylang/module.py [hunk 1]
```diff
@@ -169,12 +171,18 @@ def _check_defs(
...
         """Compiles the module and returns the final Hugr."""
         if self.compiled:
             assert self._compiled_hugr is not None, "Module is compiled but has no Hugr"
             return self._compiled_hugr
 
+        # Prepare Hugr for this module
+        graph = Module()
+        # TODO: Metadata not yet supported
+        # https://github.com/CQCL/hugr/issues/1319
+        # graph.set_root_metadata("name", self.name)
```
Comment (aborgna-q) [inline guppylang/module.py:184]: https://github.com/CQCL/hugr/issues/1319

File: guppylang/prelude/_internal/checker.py
Comment (aborgna-q) [inline guppylang/prelude/_internal/checker.py:1]: Split `_internal` into multiple files

File: pyproject.toml [hunk 2]
```diff
@@ -13,7 +13,8 @@ graphviz = "^0.20.1"
 networkx = "^3.2.1"
 pydantic = "^2.7.0b1"
 typing-extensions = "^4.9.0"
-hugr = "^0.5.0"
+#hugr = "^0.6.0"
+hugr = { git = "https://github.com/CQCL/hugr", branch = "ab/guppy-things", subdirectory = "hugr-py" }
```
Comment (aborgna-q) [inline pyproject.toml:17]: Waiting for `hugr 0.6.1`, as we need the fix from https://github.com/CQCL/hugr/pull/1422
We'll also need a tket2 release compatible with it

File: tests/integration/test_tket.py [hunk 3]
```diff
@@ -25,7 +25,9 @@
 tket2_installed = find_spec("tket2") is not None
 
 
-@pytest.mark.skipif(not tket2_installed, reason="Tket2 is not installed")
+# TODO: Remove this skip once https://github.com/CQCL/hugr/issues/1424 is fixed
```
Comment (aborgna-q) [inline tests/integration/test_tket.py:28]: The bug from https://github.com/CQCL/hugr/issues/1424 causes tket2 to not recognise the float consts

Comment (aborgna-q) [reply tests/integration/test_tket.py]: Fixed in https://github.com/CQCL/hugr/pull/1427

File: guppylang/prelude/_internal/util.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+            Any other types is treated as a native python type, and translated to
+            a Hugr type using `builtin_type_to_hugr`.
+        args: The type arguments of the operation.
+        ext: The extension of the operation. Defaults to a placeholder extension.
+        variable_remap: A mapping from the hugr param variable indices to
+            de Bruijn indices in the guppy type. Defaults to identity.
+
+    Returns:
+        A function that takes an instantiation of the type arguments and returns
+        a concrete HUGR op.
```
Comment (mark-koch) [inline guppylang/prelude/_internal/util.py:263]: I think it would be better to take the whole (monomorphic) signature instead of the type args. `CustomCallCompiler.compile_call` can should be able to provide this signature.

Then you could get rid of most of the hacky things in this file

File: guppylang/prelude/builtins.py [hunk 5]
```diff
@@ -69,25 +80,27 @@ class array(Generic[_T, _n]):
...
 class Bool:
-    @guppy.hugr_op(builtins, logic_op("And", [tys.TypeArg(tys.BoundedNatArg(n=2))]))
+    @guppy.hugr_op(builtins, logic_op("And", 2))
     def __and__(self: bool, other: bool) -> bool: ...
 
     @guppy.custom(builtins, NoopCompiler())
     def __bool__(self: bool) -> bool: ...
 
-    @guppy.hugr_op(builtins, int_op("ifrombool"))
+    @guppy.hugr_op(builtins, int_op("ifrombool", [bool], [int], n_vars=0))
```
Comment (mark-koch) [inline guppylang/prelude/builtins.py:89]: If you go with my signature suggestion from above, you wouldn't need to duplicate the signature in the decorator.

Also, I'm not a big fan of using Python types here...

**Review by mark-koch — COMMENTED** [review]
Thank you, this was a lot of work!

Generally, things look very good, the builder makes the compilation code a lot cleaner 👍

Most comments below are just nits. The thing I'm least sure about are the hacks for the builtins (see my suggestion of passing the signature instead of only type args to the `op` function).

File: examples/demo.ipynb
[the remark this answers is not in the document]
Comment (aborgna-q) [reply examples/demo.ipynb:1]: Fixed. Not sure where did that come from

File: guppylang/compiler/core.py [hunk 6]
```diff
@@ -1,14 +1,17 @@
 from abc import ABC
 from dataclasses import dataclass, field
+from typing import cast
+
+from hugr import Wire, ops
+from hugr.dfg import DP, _DfBase
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/core.py:6]: Added an issue in hugr.
https://github.com/CQCL/hugr/issues/1441

We can update the code here once that is published.

File: guppylang/compiler/core.py [hunk 7]
```diff
@@ -21,11 +24,19 @@ class DFContainer:
...
-    graph: Hugr
-    node: DFContainingNode
+    builder: _DfBase[ops.DfParentOp]
     locals: CompiledLocals = field(default_factory=dict)
 
-    def __getitem__(self, place: Place) -> OutPortV:
+    def __init__(
+        self, builder: _DfBase[DP], locals: CompiledLocals | None = None
+    ) -> None:
+        generic_builder = cast(_DfBase[ops.DfParentOp], builder)
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/core.py:33]: Added that to https://github.com/CQCL/hugr/issues/1441

File: guppylang/compiler/core.py [hunk 8]
```diff
@@ -1,14 +1,17 @@
...
 
 from guppylang.checker.core import FieldAccess, Place, PlaceId, Variable
 from guppylang.definition.common import CompiledDef, DefId
 from guppylang.error import InternalGuppyError
-from guppylang.hugr_builder.hugr import DFContainingNode, Hugr, OutPortV
 from guppylang.tys.ty import StructType
 
 CompiledGlobals = dict[DefId, CompiledDef]
-CompiledLocals = dict[PlaceId, OutPortV]
+CompiledLocals = dict[PlaceId, Wire]
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/core.py:14]: `Wire` is the protocol identifying edges. It's implemented by ports, nodes, and the graph builders.
`CompiledLocals` should not need to care about the concrete classes.

File: guppylang/compiler/expr_compiler.py [hunk 9]
```diff
@@ -341,37 +382,44 @@ def visit_ResultExpr(self, node: ResultExpr) -> OutPortV:
...
             if not gens:
                 list_port, elt_port = self.visit(list_name), self.visit(elt)
-                push = self.graph.add_node(
-                    DummyOp("Push"), inputs=[list_port, elt_port]
+                elt_ty = get_type(elt)
+                push = self.builder.add_op(
+                    list_push_op(list_ty, elt_ty), list_port, elt_port
                 )
-                self.dfg[list_place] = push.add_out_port(list_port.ty)
+                self.dfg[list_place] = push
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/expr_compiler.py:422]: Yep, we can use the node directly here.

File: guppylang/tys/ty.py [hunk 10]
```diff
@@ -276,32 +275,28 @@ def cast(self) -> "Type":
...
-                        extension="arithmetic.int.types",
-                        id="int",
-                        args=[tys.TypeArg(tys.BoundedNatArg(n=NumericType.INT_WIDTH))],
-                        bound=tys.TypeBound.Copyable,
-                    )
+                return ht.Opaque(
+                    extension="arithmetic.int.types",
+                    id="int",
+                    args=[ht.BoundedNatArg(n=NumericType.INT_WIDTH)],
+                    bound=ht.TypeBound.Copyable,
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/tys/ty.py]: We're now using `hugr.std.{int,float}`

File: guppylang/compiler/expr_compiler.py [hunk 11]
```diff
@@ -72,95 +68,104 @@ def compile_row(self, expr: ast.expr, dfg: DFContainer) -> list[OutPortV]:
...
         assert len({inp.place.id for inp in inputs}) == len(
             inputs
         ), "Inputs are not unique"
-        self.dfg = DFContainer(self.graph, node, self.dfg.locals.copy())
-        for input_node in inputs:
-            self.dfg[input_node.place] = inp.add_out_port(input_node.place.ty)
-        with self.graph.parent(node):
-            yield
+        self.dfg = DFContainer(builder, self.dfg.locals.copy())
+        hugr_input: Node = builder.input_node
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/expr_compiler.py:90]: dropped the annotation

File: guppylang/compiler/expr_compiler.py [hunk 12]
```diff
@@ -72,95 +68,104 @@ def compile_row(self, expr: ast.expr, dfg: DFContainer) -> list[OutPortV]:
...
                 # Note that we have to do fresh calls to `self.visit` here since we're
                 # in a new context
-                [self.visit(branch), *(self.visit(name) for name in loop_vars)]
+                self.visit(branch),
+                *(self.visit(name) for name in loop_vars),
             )
         # Update the DFG with the outputs from the loop
-        for node in loop_vars:
-            self.dfg[node.place] = loop.add_out_port(node.place.ty)
+        for node, wire in zip(loop_vars, loop.parent_node, strict=True):
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/expr_compiler.py:121]: It wasn't needed. I've removed it (and the other appearance further down).

File: guppylang/compiler/expr_compiler.py [hunk 13]
```diff
@@ -282,39 +326,36 @@ def visit_TypeApply(self, node: TypeApply) -> OutPortV:
...
         # since it is not implemented via a dunder method
         if isinstance(node.op, ast.Not):
+            from guppylang.prelude._internal.util import logic_op
+
             arg = self.visit(node.operand)
-            op = ops.CustomOp(extension="logic", name="Not", args=[], parent=UNDEFINED)
-            return self.graph.add_node(ops.OpType(op), inputs=[arg]).add_out_port(
-                bool_type()
+            op = logic_op("Not", inputs=1, parametric_size=False)(
+                [TypeArg(bool_type())]
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/expr_compiler.py:331]: It was unused. I replaced the whole thing with `hugr.std.logic.Not`.

File: guppylang/compiler/expr_compiler.py [hunk 14]
```diff
@@ -282,39 +326,36 @@ def visit_TypeApply(self, node: TypeApply) -> OutPortV:
...
+            return self.builder.add_op(op, arg)
 
         raise InternalGuppyError("Node should have been removed during type checking.")
 
-    def visit_FieldAccessAndDrop(self, node: FieldAccessAndDrop) -> OutPortV:
+    def visit_FieldAccessAndDrop(self, node: FieldAccessAndDrop) -> Wire:
         struct_port = self.visit(node.value)
-        unpack = self.graph.add_unpack_tuple(struct_port)
-        return unpack.out_port(node.struct_ty.fields.index(node.field))
+        return self._unpack_tuple(struct_port, [f.ty for f in node.struct_ty.fields])[0]
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/expr_compiler.py:336]: Wow, good catch!
I added a test in `test_struct.py`

File: guppylang/tys/ty.py [hunk 15]
```diff
@@ -146,21 +145,21 @@ class BoundTypeVar(TypeBase, BoundVar):
...
         # This is fine since Guppy doesn't use the equatable feature anyways.
-        return TypeBound.Copyable
+        return ht.TypeBound.Copyable
 
     def cast(self) -> "Type":
         """Casts an implementor of `TypeBase` into a `Type`."""
         return self
 
-    def to_hugr(self) -> tys.Type:
+    def to_hugr(self) -> ht.Type:
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/tys/ty.py]: Done!

File: guppylang/compiler/func_compiler.py [hunk 16]
```diff
@@ -1,66 +1,73 @@
...
     func: CheckedNestedFunctionDef,
     dfg: DFContainer,
-    graph: Hugr,
     globals: CompiledGlobals,
-) -> OutPortV:
-    """Compiles a local (nested) function definition to Hugr."""
+) -> tuple[Function, Wire]:
+    """Compiles a local (nested) function definition to Hugr and loads it into a value.
+
+    Returns the compiled function node, and the wire output of the `LoadFunc` operation.
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/func_compiler.py:35]: It's not currently used. I imagined it'd be useful to keep the function builder around for recursive calls and the like, but it wasn't needed.
I'll remove it.

File: guppylang/compiler/func_compiler.py [hunk 17]
```diff
@@ -1,66 +1,73 @@
...
     )
 
     # If we have captured variables and the body contains a recursive occurrence of
     # the function itself, then we provide the partially applied function as a local
     # variable
-    if len(captured) > 0 and func.name in func.cfg.live_before[func.cfg.entry_bb]:
-        loaded = graph.add_load_function(def_node.out_port(0), [], def_node).out_port(0)
-        partial = graph.add_partial(
-            loaded, [def_input.out_port(i) for i in range(len(captured))], def_node
+    call_args = cast(list[Wire], func_builder.inputs())
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/compiler/func_compiler.py:61]: Nice catch. I'll just copy the list, I think using `Wire` is cleaner than extracting ports from the builders all the time.

File: guppylang/definition/custom.py [hunk 18]
```diff
@@ -156,35 +184,36 @@ def load_with_args(
...
-                type_args,
-                DFContainer(graph, def_node),
-                graph,
-                globals,
-                node,
-            )
-            graph.add_output(returns)
+        #
+        # TODO: Reuse compiled instances with the same type args?
+        # TODO: Why do we need to monomorphise here?
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/definition/custom.py:180]: Added that to the definition-reuse issue, since they can be done together.

File: guppylang/definition/custom.py [hunk 19]
```diff
@@ -156,35 +184,36 @@ def load_with_args(
...
+        fun_ty: FunctionType = self.ty.instantiate(type_args)
+        input_types: list[ht.Type] = [ty.to_hugr() for ty in fun_ty.inputs]
+        output_types: list[ht.Type] = [fun_ty.output.to_hugr()]
+        func: hf.Function = dfg.builder.define_function(
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/definition/custom.py:184]: Because I like typing stuff :P
I dropped the annotations

File: guppylang/definition/struct.py [hunk 20]
```diff
@@ -194,15 +195,14 @@ def check_instantiate(
         check_all_args(self.params, args, self.name, loc)
         return StructType(args, self)
 
-    @cached_property
-    def generated_methods(self) -> list[CustomFunctionDef]:
+    def generated_methods(self, module: hf.Module) -> list[CustomFunctionDef]:
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/definition/struct.py]: Dropped it

File: guppylang/prelude/builtins.py (hunk 5 above)
Comment (aborgna-q) [reply guppylang/prelude/builtins.py]: Done, we still need some extra parameters to specify the required arguments (since those don't correspond to guppy op args), but the input/output rows are gone.

**Review by mark-koch — APPROVED** [review]
Nice work, I think this is ready to be merged! I'm happy with your builtins solution, way nicer than before.
